### DISCUSSION TITLE
Help | How to implement save in dokan

### DISCUSSION THREAD

**Discussion by biplab5464** [dokan-dev/dokany#1230] [Q&A]
From my understanding of the documentation, it seems that for any save event, the create_file function is called first, followed by the write_file function. However, I’m not clear on the conditions under which the write_file function is invoked.

At present, I’m able to recursively list directories and open files. However, when I attempt to save a file, I encounter an error in VS Code: Failed to save 'test/c': Unable to write file 'm:\test\test.c' (Unknown (FileSystermError): Error: EINVAL: invalid argument, write).

Despite having implemented the write_file function, I’m still facing this issue. As I’m new to Windows development and still familiarizing myself with the Windows API, I’m finding it challenging to understand everything.

I was wondering if anyone could assist me with this issue, perhaps through a platform like Discord or Telegram, instead of creating an issue here. Any help would be greatly appreciated.

**Comment by Liryna** [comment]
Hi @biplab5464 , I would suggest implementing all dokan operations. Use [Procmon](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) to investigate where does the error comes from (create, write ...) and then look at your filesystem logs to determine the error.

You can also try testing tools like [1](https://github.com/Liryna/fstools) and [2](https://github.com/Liryna/winfstest) to ensure your filesystem behaves as expected

**Comment by biplab5464** [comment]
[Logfile.CSV](https://github.com/user-attachments/files/15587781/Logfile.CSV)

when create file is called Desired access is : Read Attributes. is it the problem?

**Comment by Liryna** [comment]
if you execute the same operation on a normal NTFS filesystem, what is the difference ?
Looks like all `QueryAllInformationFile` are failing with`BUFFER OVERFLOW`.

**Comment by biplab5464** [comment]
for every process `QueryAllInformationFile` is failing with `BUFFER OVERFLOW` 

For normal NTFS some have Read Attributes and some have Generic Read and write attributes and synchronization

**Comment by biplab5464** [comment]
I also try implementing `delete_file` it is also not working, i mean i have unable to see the print statement inside `delete_file` not showing up in the console 

is there is something that this function doesn't print to the console or this function is not being called

**Comment by biplab5464** [comment]
```
                        ================>info<=============
trying to open file \build.rs\rust.rs\test.rtf
desired_access : 128
            file_attributes : 0
            sharePacces : 7
            create_disposition : 3
            create_option : 2097152
            flags_and_attributes; 2097152
            is_dir : false
desired access flags
FILE_READ_ATTRIBUTES access requested
file attribute falgs
FILE_FLAG_OPEN_REPARSE_POINT
open a fle Open existing
````
this is debug output from the create_file, when it is trying to open `\build.rs\rust.rs\test.rtf` this path while the acutal path is `M:\test\test\test.rtf`  
while i am trying to safe it is also referring to this path  but the file is opened 

![image](https://github.com/dokan-dev/dokan-rust/assets/45629823/4475d6f5-9775-415d-85a7-c0196e3a1fcb)

am i doing something wrong here

**Comment by biplab5464** [comment]
one unique thing is both `build.rs` and `rust.rs` are the first file from after dir listing end

**Comment by Liryna** [comment]
Please see
* https://dokan-dev.github.io/dokany-doc/html/
* [DeleteOnClose](https://dokan-dev.github.io/dokany-doc/html/struct_d_o_k_a_n___f_i_l_e___i_n_f_o.html#a89b3ba429fd6c98b1ee30d100f9441fb)
* https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-ntcreatefile?redirectedfrom=MSDN `FILE_DELETE_ON_CLOSE`
* It is normal to receive the path in the scope of your mount point. The [samples](https://github.com/dokan-dev/dokany/tree/master/samples/dokan_memfs) would also help you understand how the operations needs to be implemented.

**Comment by biplab5464** [comment]
Thank you for the help. I am a beginner in Windows development. I have previously worked on Linux, so a lot of this is new to me, but I am making progress.

I am now able to write to a text file and the `write_file` function is working. However, I am having problems with `docx` or `rtf` files.

When I open any file with VS Code or Notepad++, even unsupported files like docx, it opens in the correct path. But with applications like WPS or WordPad, docx or rtf files are not opening in the correct path.
![image](https://github.com/dokan-dev/dokan-rust/assets/45629823/9236bbde-69d2-4f01-b258-b7485e32419a)
` For example, the correct path should be M:\test\test.docx.`

The same issue occurs when trying to delete a file. For instance, when I click on the file test.c and try to delete it from the context menu, the prompt asks me to delete build.rs instead."
![image](https://github.com/dokan-dev/dokan-rust/assets/45629823/1af82dee-ef2a-4ae7-b0fb-83637daeaa80)

**Comment by Liryna** [comment]
Good progress!

The incorrect path must be due to an incorrect name returned during find files or get file info. Please check what your implementation returns and what the samples are returning (expectation).

**Comment by Liryna** [comment]
@biplab5464 have you been able to make progress? Can this be closed ?

**Comment by biplab5464** [comment]
i am able to make lots of progress. thank you.

i am still not able to figure out the wrong path in some apps 

like for a jpeg file, a file is opening it at the wrong path for windows photo app but at the same time it is opening correctly in vs code or other image viewer 

i have looked at `find_files` and `get_file_imformation` several times as you have suggested and tweaked it like i have tried several ways to find unique number to `file_index` but still having the issue 

and better we can close the issue and discuss this on some dokan-rust forum or discord.

if something like this exists can you please share it with me.  so we can closed this issue 

Thank you

**Comment by biplab5464** [comment]
about the wrong path it is always returning the first file in the directory with alphabetically sort

**Comment by biplab5464** [comment]
My Friend is testing my my dokan app it was working perfectly for him, i mean he didn't got wrong Path for any app or in any condition but i still getting wrong path although the code is same also the driver version is same. Also, we tried to replicate same folder structure (i send him the file i was testing on)
the only difference is our OS

my system 
Edition	Windows 11 Home Single Language
Version	23H2
Installed on	‎28-‎02-‎2024
OS build	22631.3155
Experience	Windows Feature Experience Pack 1000.22684.1000.0

My friend system 
Edition	Windows 11 Pro
Version	23H2
Installed on	‎5/‎11/‎2024
OS build	22631.3737
Experience	Windows Feature Experience Pack 1000.22700.1009.0
